---
title: "Differential Expression Analysis"
author: "Sana Badri"
date: "3/9/2021"
---


>-  Process of analyzing the differences in RNA expression between multiple samples and groups from Counts data.



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###Load DESeq2 dataset:
>- From transcript abundance files and tximport
>- From a count matrix
>- From htseq-count files
>- From a SummarizedExperiment object


```{r, echo = TRUE}
### LOAD REQUIRED LIBRARIES
library("DESeq2")
library("pheatmap")
library("RColorBrewer")
library("vsn")
library("AnnotationDbi")
library("org.Hs.eg.db")
library("genefilter")
library("biomaRt")
library("IHW")
library("ggplot2")


#setwd("/Users/sanabadri/Desktop/SPRING2021/ASI/Practicum 6 - Hands-on gene expression analysis/")
CountTable <- read.csv("kal2.csv", header=TRUE, row.names=1) 
samples <- read.table("sample.txt", header=TRUE)

dds <- DESeqDataSetFromMatrix(countData = CountTable, colData=samples, design=~batch+condition)
#dds <- DESeqDataSetFromMatrix(countData = CountTable, colData=samples, design=~condition)

dds = DESeq(dds)
counts(dds)->raw_counts
raw_counts<-as.data.frame(counts(dds))


#Not necessary to pre-filter low count genes before running the DESeq2 functions, pre-filtering is useful and can be done by removing rows in which there are very few reads, so we reduce the memory size of the dds data object, and increase the speed of the transformation and testing functions within DESeq2

#keep <- rowSums(counts(dds)) >= 10
#dds <- dds[keep,]


#Create a normalized matrix
norm_counts = counts(dds, normalized = TRUE)

### PRELIMINARY ANALYSES ###
# The first steps in your analysis should focus on better understanding the relationship of the datasets being studied. This can
# be simply achieved by generating a PCA plot showing the relationship of your samples.
# First we transform our raw count data using a variance stabilizing transformation (VST) that roughly mirrors how DeSeq2 models the data.
vsd1 <- varianceStabilizingTransformation(dds, blind=FALSE)

# Then we plot a PCA, grouping and coloring our datasets according to batch
plotPCA(vsd1, "condition")

# note that you can attach additional information based on the column headers in your sample table
plotPCA(vsd1, c("condition","batch"))

# we can also attempt to replicate the batch effect correction performed by DeSeq2 using the limma::removeBatchEffect function
vsd2 <- varianceStabilizingTransformation(dds, blind=FALSE)
assay(vsd2) <- limma::removeBatchEffect(assay(vsd2), vsd2$batch)
plotPCA(vsd2, "condition")
data <- plotPCA(vsd2, "condition",returnData=TRUE)

# We can also calculate and plot sample distances using either the batch corrected (vsd2) or uncorrected (vsd1) data. 
# uncorrected
sampleDists <- dist( t( assay(vsd1) ) )
sampleDists
sampleDistMatrix <- as.matrix( sampleDists )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)
pheatmap(sampleDistMatrix, clustering_distance_rows=sampleDists, clustering_distance_cols=sampleDists, col=colors)
# corrected
sampleDistsCorr <- dist( t( assay(vsd2) ) )
sampleDistsCorr
sampleDistCorrMatrix <- as.matrix( sampleDistsCorr )
colnames(sampleDistCorrMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistCorrMatrix, clustering_distance_rows=sampleDists, clustering_distance_cols=sampleDists, col=colors)




data <- plotPCA(vsd2, returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))
ggplot(data, aes(PC1, PC2, color=condition)) +
  theme_bw() + 
  theme(panel.background = element_blank())+
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))+
  geom_point(size=5)+
  scale_color_manual(values=c("#74c476","#ec7014","purple"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  theme(plot.title = element_blank())+
  theme(text = element_text(size=20))+
  theme(legend.position="bottom") +
  theme(axis.text = element_text( color = "black", size = 25))+
  theme(legend.title=element_blank())

```

# At this stage, you should have a good sense of how your samples cluster and the effect of batch correction (if used)
# In simple RNA-Seq situations (control vs treatment, 3-5 bioreps each), this is all that should be required. 
# For more complex situations, you will need to dive deep into working of DeSeq2
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html


```{r BASIC DGE ANALYSIS USING DESEQ2}

# Run DESEQ and generate a simple plot showing the distribution of regulated and unregulated genes
DatasetProcessed <- DESeq(dds) # runs DESEQ
par(mfrow=c(1,1))
#plotMA(DatasetProcessed, main="DESeq2", ylim=c(-5,5))

# Next we perform a contrast analysis to produce a list of differentially regulated genes between our two conditions
# First we set CTRL dataset as baseline
DatasetProcessed$condition <- relevel(DatasetProcessed$condition, "Ctrl")

# Next we create our results object while performing shrinkage of effect size 
# (this reduces the impact of apparent gross changes in low expressed genes)
res1 <- lfcShrink(DatasetProcessed, coef=2)



# Here we modify our output data to include two additional columns that contain the baseMeans (a proxy for counts)
# This is useful for downstream filtering of lowly expressed genes
baseMeanCtrl = rowMeans(counts(DatasetProcessed,normalized=TRUE)[,DatasetProcessed$condition == "Ctrl"])
baseMeansiRNA = rowMeans(counts(DatasetProcessed,normalized=TRUE)[,DatasetProcessed$condition == "siRNA"])
res1 = cbind(as.data.frame(res1), baseMeanCtrl, baseMeansiRNA)

# Here we add two further columns, the gene symbol (common name) and entrez ID - both of which may be useful downstream
res1$symbol <- mapIds(org.Hs.eg.db, keys=row.names(res1), column="SYMBOL", keytype="ENSEMBL", multiVals="first") # MAPS GENE IDs
res1$entrez <- mapIds(org.Hs.eg.db, keys=row.names(res1), column="ENTREZID", keytype="ENSEMBL", multiVals="first")

# Finally we write the complete results object to an outfile
write.csv(res1, "DGEanalysis.csv", row.names=TRUE)
```



## Differential analysis
```{r DE analysis with DESEQ2}


Ctrl_siRNA<-res1
head(`Ctrl_siRNA`)
dim(`Ctrl_siRNA`)

adj_p_val = 0.05
abs_log2fc = 1

Ctrl_siRNA_sig = res1[res1$padj < adj_p_val & !is.na(res1$padj) & abs(res1$log2FoldChange) >abs_log2fc,]
head(`Ctrl_siRNA_sig`)
dim(`Ctrl_siRNA_sig`)



ds2_df<-as.data.frame(`Ctrl_siRNA`)
ds2_df$log10.pvalue<-(-1*log10(ds2_df$padj))
ds2_df$Significant.Gene<-"No"

row_number<-which(ds2_df$padj<adj_p_val & abs(ds2_df$log2FoldChange)>abs_log2fc)
ds2_df$Significant.Gene[row_number]<-"Yes"
ds2_df$Differential_Gene<-ds2_df$Significant.Gene
ds2_df$Differential_Gene[which(ds2_df$padj< adj_p_val & ds2_df$log2FoldChange> abs_log2fc)]<-"Up-Regulated"
ds2_df$Differential_Gene[which(ds2_df$padj< adj_p_val & ds2_df$log2FoldChange< -abs_log2fc)]<-"Down-Regulated"

sub_ds2 = subset(ds2_df, abs(ds2_df$log2FoldChange) > 1 & ds2_df$padj < .05)

write.csv(ds2_df, "ds2.csv", row.names=TRUE)
write.csv(sub_ds2, "sub_ds2.csv", row.names=TRUE)
de1 <- decideTestsDGE(et12, adjust.method="BH", p.value=0.05)

```

```{r Volcano plot with ggplot2}
library(ggplot2)
library(ggrepel)
#RNA volcano

#make sure to adjust xlim and ylim based off of the data. Run hist(ds2_df$log2FoldChange) and hist(ds2_df$log10.pvalue) to get an idea

# pick a set of genes to label:
ds2_df$label<-"No"
ds2_df$label[which(ds2_df$padj<adj_p_val & abs(ds2_df$log2FoldChange)>abs_log2fc)]<-"Yes"

# Here we add two further columns, the gene symbol (common name) and entrez ID - both of which may be useful downstream
ds2_df$symbol <- mapIds(org.Hs.eg.db, keys=row.names(ds2_df), column="SYMBOL", keytype="ENSEMBL", multiVals="first") # MAPS GENE IDs
ds2_df$entrez <- mapIds(org.Hs.eg.db, keys=row.names(ds2_df), column="ENTREZID", keytype="ENSEMBL", multiVals="first")

#pdf("Volcano.pdf", width=8, height=4,useDingbats=FALSE)
pRNA <- ggplot(ds2_df, aes(log2FoldChange, log10.pvalue))+
  geom_point(aes(colour = Differential_Gene),size=1)+
  scale_colour_manual(values=c("blue", "black","red"))+
  xlim(-6,6)+
  ylim(0,5)+
  xlab(expression(paste(log[2], 'FC(Ctrl/siRNA)')))+
  ylab(expression(paste(-log[10], '(Q-value)'))) +
  theme_bw() + 
  theme(panel.background = element_blank())+
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1))+

  theme(text = element_text(size=10)) +
  theme(plot.title = element_blank())+
  theme(axis.text = element_text( color = "black", size = 10))+
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.line=element_line()) +
  theme(legend.position="none")+
  geom_vline(xintercept=-abs_log2fc,linetype="dotted")+
  geom_vline(xintercept=abs_log2fc,linetype="dotted")+
  geom_text_repel(
    data = subset(ds2_df, label=="Yes"),
    aes(label = ds2_df$symbol[which(label=="Yes")]),
    size = 5,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.5, "lines")
  )
pRNA

#dev.off()

```


## Heatmap with DEGs
```{r Heatmap}
sig.genes<-rownames(`Ctrl_siRNA_sig`)

#extract raw counts for these genes:
m<-match(sig.genes,rownames(norm_counts))
de_norm_counts<-norm_counts[m,]


library(pheatmap)
library(RColorBrewer)

sample<-as.data.frame(data$name)
colnames(sample)<-c("samples")
rownames(sample)<-colnames(de_norm_counts)
sample$samples<-as.character(data$condition)
sample2<-sample
sample2$samples[which(sample2$samples=="siRNA")]<-"#74c476"
sample2$samples[which(sample2$samples=="Ctrl")]<-"#ec7014"
sample3 = list(samples = c("siRNA"= "#74c476", "Ctrl"="#ec7014"))

#pdf("HEATMAP.pdf", width=10, height=5,useDingbats=FALSE)
pheatmap(de_norm_counts,scale = "row",show_rownames=FALSE,
         colorRampPalette(rev(brewer.pal(n = 10, name ="RdBu")))(25),
         clustering_method = "ward.D2",clustering_distance_cols ="euclidean",
         annotation_col  = sample)
#dev.off()

```

